### ISSUE TITLE
ESQL: Inconsistent/buggy behavior for stats with null expression

### ISSUE THREAD

**Issue by alex-spies** [elastic/elasticsearch#104900]
Given an index `testidx` with, let's say, 5 documents in it, `from testidx | STATS some_agg_function(null)` behaves differently for different aggregations:

- `from testidx | STATS count(null)` returns the number of documents (like `count(*)`).
- `from testidx | STATS avg(null)` returns a row containing `null` for every document.
- `from testidx | STATS max(null)` returns a 500 (`{"type":"esql_illegal_argument_exception","reason":"unknown agg: class org.elasticsearch.xpack.ql.expression.Literal: null"}`)

**Comment by alex-spies** [comment]
I think it shouldn't make a difference if there's a `null` literal or if the null comes from an eval; i.e. `count(null)` should be equivalent to `eval x = null | stats count (x)`. This is currently not the case, the latter returns `0`.

**Comment by costin** [comment]
count(field) only considers non null values.
count(literal) is a special case of count(*) - for every entry in the storage count looks at the argument - the literal, which is not null hence why all rows are returned since none of them are null; they get replaced by the literal.
count(null) however can only return empty (if there's no data) or 0 since even if all the data is non-null, count is instructed to count each entry as null, which means it drops it.

**Comment by alex-spies** [comment]
> > from testidx | STATS count(null) returns the number of documents (like count(*)).
> 
> That is incorrect.
> 
> count(field) only considers non null values. count(literal) is a special case of count(*) - for every entry in the storage count looks at the argument - the literal, which is not null hence why all rows are returned since none of them are null; they get replaced by the literal. count(null) however can only return empty (if there's no data) or 0 since even if all the data is non-null, count is instructed to count each entry as null, which means it drops it.

I agree that this is the intended behavior - `count(null)` should be either `0` or `null`. I just ran this on main, however, and current behavior seems still to be that `count(null) == count(*)`.

On whether `count(null)` should be 0 or `null`: I compared with MariaDB and PostgreSQL, both of which return `0`.
Our own `_sql` returns `null`, both for `count(null)` but also for `avg(null)` and `sum(null)`. (Postgres just errors out for `avg(null)` and `sum(null)`).

> 
> > from testidx | STATS avg(null) returns a row containing null for every document.
> 
> I would expect null to be returned since avg(null)=sum(null)/count(null)-> null/null-> null
Yes, that makes sense. Current behavior is buggy though, and has one row with `null` per document in the index if `avg(null)` is the only aggregation in the stats command.
